### PR TITLE
COMP: don't complete unstable stdlib items when on stable or beta

### LINKED ISSUE #4437: Do not propose unstable features when using stable rustc
## Environment

* **IntelliJ Rust plugin version:** 0.2.106.2134-193
* **Rust toolchain version:** 1.37.0 (eae3437df 2019-08-13) x86_64-apple-darwin
* **IDE name and version:** CLion 2019.3 EAP (CL-193.3519.24)
* **Operating system:** macOS 10.14.6

## Problem description

I have the stable rust toolchain used, but I get the completion for things as `Option::flatten` which is not available for the current stable rustc:

image

If not covered yet, other `cfg`-related completions can be also disabled when are not applicable.

### LINKED ISSUE #8836: Feature request: hide all unstable items from autocomplete
There's nothing more saddening than autocompleting a method name or trait that does *exactly what you want*, only to find out that it's been unstable since 2016, requiring you to undo and try again and possibly write the function yourself (but with extra annoyance now).

I would prefer to just not see these unstable features that I can't use. I want to be able to completely disable them showing up in autocomplete, as if they just weren't there (as I can't touch them anyway).

That way I can easily find a stable solution to the problem without possibly multiple unstable solutions not working out, causing frustration. That's because there's no way to tell stable from unstable items in autocomplete, so I can't easily tell which methods I'm allowed to use to solve my problem.

Differentiating unstable vs stable items in autocomplete would provide an earlier warning but it would stay just as annoying (that perfect method that solves your problem in 1 call is just taunting you with "instability"). Ideally there should be both; you should be able to either differentiate them, or turn unstable items off completely. But as I'd be doing the latter I only care about that.

p.s. My toolchain *is* nightly, but that doesn't matter for consumers of my libraries. Requiring nightly would just ban everyone from using it.

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes #8836, fixes #4437, 

changelog: Don't show unstable stdlib items in completion list when using stable or beta Rust compiler

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/resolve/Processors.kt [hunk 1]
```diff
@@ -511,3 +514,41 @@ fun filterDeriveProcMacros(processor: RsResolveProcessor): RsResolveProcessor =
...
+                    return ownerStability
+                }
+                val superItem = superItem
+                if (superItem != null) {
+                    return superItem.queryAttributes.stability
+                }
+            }
+        }
+
+        return null
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/resolve/Processors.kt:553]: Looks like modules by default are `#[stable]`, should we handle them separately here?

```rust
// stdlib
pub mod inner {
    #[stable]
    pub fn func() {}
}

// main.rs
fn main() {
    // currently no completion for inner
    inn/*caret*/::func();
}
```

File: src/main/kotlin/org/rust/lang/core/resolve/Processors.kt [hunk 2]
```diff
@@ -511,3 +514,41 @@ fun filterDeriveProcMacros(processor: RsResolveProcessor): RsResolveProcessor =
         if (!function.isCustomDeriveProcMacroDef) return@wrapWithFilter false
         true
     }
+
+fun RsOuterAttributeOwner.shouldHideElementInCompletion(contextMod: RsMod): Boolean {
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/resolve/Processors.kt:518]: Btw interesting fact that there can be stable item inside unstable module - `core::unicode::UNICODE_VERSION`

File: src/main/kotlin/org/rust/lang/core/completion/RsCommonCompletionProvider.kt [hunk 3]
```diff
@@ -207,8 +207,14 @@ object RsCommonCompletionProvider : RsCompletionProvider() {
...
+
         for (candidate in candidates) {
             val item = candidate.item
+            if (item is RsOuterAttributeOwner) {
+                val isHidden = item.shouldHideElementInCompletion(contextMod)
+                if (isHidden) continue
+            }
             val scopeEntry = SimpleScopeEntry(candidate.itemName, item)
 
             if (item is RsEnumItem
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/completion/RsCommonCompletionProvider.kt:220]: Fixed!
I've tested it on `std::io::ErrorKind` enum

**Review by dima74 — APPROVED** [review]
bors r+

**Comment by vlad20012** [comment]
bors retry

**Comment by Undin** [comment]
BTW Java uses a different approach with items not available with current language level. Examples below for java 11 target

image

Screenshot 2023-02-21 at 11 08 25

**Pull request merged**